File: data_structures/trie/python/test_trie.py
from trie import Trie


def test_new_size_zero():
    t = Trie()
    assert t.size() == 0


def test_insert_and_search():
    t = Trie()
    t.insert("apple")
    assert t.search("apple")
    assert not t.search("app")
    assert t.starts_with("app")
    assert not t.starts_with("ban")
    assert t.size() == 1


def test_shared_prefix():
    t = Trie()
    t.insert("apple")
    t.insert("app")
    assert t.search("app")
    assert t.search("apple")
    assert t.size() == 2
    assert t.remove("app")
    assert t.search("apple")
    assert not t.search("app")
    assert t.size() == 1


def test_multiple_words():
    t = Trie()
    t.insert("cat")
    t.insert("car")
    t.insert("card")
    assert t.starts_with("ca")
    assert t.starts_with("car")
    assert not t.starts_with("cb")
    assert not t.remove("cap")


def test_remove_absent():
    t = Trie()
    t.insert("hello")
    assert not t.remove("hell")
    assert t.size() == 1

File: data_structures/trie/matlab/test_trie.m
% test_trie.m

t = trie('new');
assert(trie(t, 'size') == 0);

t = trie(t, 'insert', 'apple');
assert(trie(t, 'search', 'apple'));
assert(~trie(t, 'search', 'app'));
assert(trie(t, 'starts_with', 'app'));
assert(~trie(t, 'starts_with', 'ban'));
assert(trie(t, 'size') == 1);

t = trie(t, 'insert', 'app');
assert(trie(t, 'search', 'app'));
assert(trie(t, 'search', 'apple'));
assert(trie(t, 'size') == 2);
[t, ok] = trie(t, 'remove', 'app');
assert(ok);
assert(trie(t, 'search', 'apple'));
assert(~trie(t, 'search', 'app'));
assert(trie(t, 'size') == 1);

t2 = trie('new');
t2 = trie(t2, 'insert', 'cat');
t2 = trie(t2, 'insert', 'car');
t2 = trie(t2, 'insert', 'card');
assert(trie(t2, 'starts_with', 'ca'));
assert(trie(t2, 'starts_with', 'car'));
assert(~trie(t2, 'starts_with', 'cb'));
[~, ok] = trie(t2, 'remove', 'cap');
assert(~ok);

disp('All tests passed.');
